pallets/pallet-bonded-coins/src/curves/mod.rs

```diff
@@ -111,6 +111,31 @@ pub(crate) fn convert_to_fixed<T: Config>(
 	let decimals = 10u128
 		.checked_pow(u32::from(denomination))
 		.ok_or(ArithmeticError::Overflow)?;
-	let scaled_x = x.checked_div(decimals).ok_or(ArithmeticError::DivisionByZero)?;
-	scaled_x.checked_to_fixed().ok_or(ArithmeticError::Overflow)
+
+	// Divide `x` to separate the quotient and remainder
+	let quotient = x.checked_div(decimals).ok_or(ArithmeticError::DivisionByZero)?;
+	let remainder = x.checked_rem_euclid(decimals).ok_or(ArithmeticError::DivisionByZero)?;
+
+	// Convert quotient to fixed-point
+	// This can overflow if capacity(u128) / 10^denomination > capacity(fixed)
+	let fixed_quotient: CurveParameterTypeOf<T> = quotient.checked_to_fixed().ok_or(ArithmeticError::Overflow)?;
+
+	// we can skip handling the remainder if it's 0
+	if remainder == 0 {
+		Ok(fixed_quotient)
+	} else {
+		// Convert remainder to fixed-point and scale it down by `decimals`
+		let fixed_remainder = remainder
+			// This can overflow if 10^denomination > capacity(float)
+			.checked_to_fixed::<CurveParameterTypeOf<T>>()
+			.ok_or(ArithmeticError::Overflow)?
+			// This WILL overflow if 10^denomination > capacity(float)
+			.checked_div(decimals.checked_to_fixed().ok_or(ArithmeticError::Overflow)?)
```

Comment:
The solution comes at the cost of one additional case where overflow can happen:

- Previously, overflow could happen where `capacity(balance) / 10^denomination > capacity(fixed)`, which for a u128 balance (capacity ~3.4e+38) and a I75F53 (capacity ~1.8e+22) meant that any denomination >= 17 is safe.
- With the current solution, overflow __will__ occur where `10^denomination > capacity(float)`, which for an I75F53 means that we have a hard cap at a denomination of 22.

We can possibly soften the hard cap, e.g. by iteratively dividing the remainder by 10 (or some other multiplicative component of `decimals` which can be represented), but denominations larger than the soft cap would progressively increase the probability of an overflow occurring. 

For example, with an I75F53 and a denomination of 23, any amount `~1.8e+22 < x < e+23` would result in a remainder that cannot be represented and thus in an overflow error.

Reply:
Softened the cap on https://github.com/KILTprotocol/kilt-node/pull/794/commits/694334c178b2082d19b0f1e27a703585b9f865ae, which I think is a better way of scaling a fixed by an integer.
Still, I introduced a constant to limit allowable denominations on e94104e29ba16db3f7626488de705a92c00068b2.

Reply:
Since 08b71c4c8966febb0c17df070b7b3e941f77115b we no longer have new cases of overflow introduced with this fix.